### PR TITLE
Make `docker pull` detect plugin content and error out.

### PR DESCRIPTION
**Pull request by vieux**
ping @anusha-ragunathan 

can you make sure I used the right media type ?

### CODE REVIEW

File: plugin/distribution/push.go [hunk 1]
```diff
@@ -79,9 +79,9 @@ func Push(name string, rs registry.Service, metaHeader http.Header, authConfig *
 			return "", err
 		}
 		f.Close()
-		mt := MediaTypeLayer
+		mt := schema2.MediaTypeLayer
```
Comment (anusha-ragunathan) [inline plugin/distribution/push.go]: Changes to this file are not needed.

File: plugin/distribution/types.go [hunk 2]
```diff
@@ -10,10 +10,5 @@ var ErrUnsupportedRegistry = errors.New("only V2 repositories are supported for
...
 var ErrUnsupportedMediaType = errors.New("content is not a plugin")
 
-// Plugin related media types
-const (
-	MediaTypeManifest = "application/vnd.docker.distribution.manifest.v2+json"
-	MediaTypeConfig   = "application/vnd.docker.plugin.v0+json"
-	MediaTypeLayer    = "application/vnd.docker.image.rootfs.diff.tar.gzip"
-	DefaultTag        = "latest"
-)
+// DefaultTag is the default tag for plugins
```
Comment (anusha-ragunathan) [inline plugin/distribution/types.go]: Changes to this file are not needed.

File: distribution/pull_v2.go [hunk 3]
```diff
@@ -356,6 +360,12 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named) (tagUpdat
 		return false, fmt.Errorf("image manifest does not exist for tag or digest %q", tagOrDigest)
 	}
 
+	if m, ok := manifest.(*schema2.DeserializedManifest); ok {
+		if m.Manifest.Config.MediaType == "application/vnd.docker.plugin.v0+json" {
```
Comment (anusha-ragunathan) [inline distribution/pull_v2.go]: Yes, `"application/vnd.docker.plugin.v0+json"` is indeed the right mediaType that image pull should check and error out.

Comment (anusha-ragunathan) [inline distribution/pull_v2.go]: @aaronlehmann : 
`application/vnd.docker.plugin.v0+json` is strictly for plugins pushed by 1.12 experimental.
What's being removed in #29230 is a temporary mediaType (which ofcourse wont be supported) we had for plugins in 1.13. This is now being permanently replaced by `application/vnd.docker.plugin.v1+json`

Comment (aaronlehmann) [inline distribution/pull_v2.go]: Just confused about why we would add something relating to the temporary media type in a patch release, when references to that media type have been removed from the upcoming version. If `application/vnd.docker.plugin.v0+json` is indeed rare, shouldn't 1.12.x warn on an attempt to pull a plugin pushed by >= 1.13.x?

Comment (anusha-ragunathan) [inline distribution/pull_v2.go]: Sounds reasonable to me.

**Comment by vieux** [comment]
@dmcgowan I updated.

**Comment by vieux** [comment]
@anusha-ragunathan tested manually works as expected:

`docker pull vieux/sshfs` -> `Error response from daemon: target is a plugin - Use `docker plugin install`

`docker pull tiborvass/no-remove` -> `Error response from daemon: target is a plugin - Use `docker plugin install`

`docker plugin install tiborvass/no-remove` -> works.

**Comment by anusha-ragunathan** [comment]
The only problem is that `docker pull vieux/sshfs -> Error response from daemon: target is a plugin - Usedocker plugin install`` is not true. You cannot install a 1.13 plugin using 1.12 experimental and expect it to work. 

So updating the error message to simply say `target is a plugin` should be good.

**Comment by vieux** [comment]
@anusha-ragunathan updated.

**Comment by anusha-ragunathan** [comment]
LGTM (granted tests pass)

File: plugin/distribution/pull.go [hunk 4]
```diff
@@ -143,8 +143,7 @@ func Pull(name string, rs registry.Service, metaheader http.Header, authConfig *
 		logrus.Debugf("pull.go: error in json.Unmarshal(): %v", err)
 		return nil, err
 	}
-	if m.Config.MediaType != MediaTypeConfig &&
-		m.Config.MediaType != "application/vnd.docker.plugin.image.v0+json" {
+	if m.Config.MediaType != MediaTypeConfig {
```
Comment (dmcgowan) [inline plugin/distribution/pull.go]: was `application/vnd.docker.plugin.image.v0+json` intended as a placeholder for `application/vnd.docker.plugin.v1+json`? What happens when `application/vnd.docker.plugin.v1+json` is encountered?

Comment (anusha-ragunathan) [inline plugin/distribution/pull.go]: Yes, `application/vnd.docker.plugin.image.v0+json` was a placeholder for the final type that 1.13 would ship with. 
If 1.12.x encounters a plugin with mediatype `application/vnd.docker.plugin.v1+json`, it implies that the plugin was created using 1.13+. Hence it is not supported in 1.12.x

**Comment by dmcgowan** [comment]
LGTM

File: plugin/distribution/pull.go (hunk 4 above)
Comment (dmcgowan) [inline plugin/distribution/pull.go]: thanks @anusha-ragunathan for explaining, the `MediaTypeConfig` here is  `application/vnd.docker.plugin.v0+json` and if anything else is encountered this condition is hit, including the 1.13 type :smile:

**Pull request merged**